src/filters/BallFilter.cpp

```diff
@@ -137,3 +161,52 @@ bool BallFilter::ballIsVisible() const {
     //If we extrapolated the ball for longer than 0.05 seconds we mark it not visible
     return (lastPredictTime - lastUpdateTime) < 0.05;
 }
+
+double BallFilter::calculateVelocity() const {
+    double xVel = kalman->state()[2];
+    double yVel = kalman->state()[3];
+    return sqrt(xVel * xVel + yVel * yVel);
+}
+
+void BallFilter::updateState()  {
+    //TODO: Tune values
+    const double thresholdRest = 0.05; // Velocity at which ball is considered resting
+    const double thresholdKick = 0.1; // Velocity at which ball is considered kicked
+    const double switchRatio = 0.67; // Phase transition point from slipping to rolling
+
+    double velocity = calculateVelocity();
+
+    switch(state) {
+        case ROLLING:
+            // When velocity is low, the ball rests
+            if (velocity < thresholdRest) {
+                state = RESTING;
+            }
+        case RESTING:
+            // Kick detection
+            // TODO: Improve kick detection
+            if (velocity - lastVelocity >= thresholdKick) {
+                state = KICKING;
+                kickVelocity = velocity;
+            }
+            break;
+        case KICKING:
+            // Update velocity during acceleration, thereafter ball starts slipping
+            if (velocity >= kickVelocity) {
+                kickVelocity = velocity;
+            } else {
+                state = SLIPPING;
+            }
+            break;
+        case SLIPPING:
+            // Ball starts rolling at a certain percentage of the kick velocity
+            if (velocity < switchRatio * kickVelocity) {
+                state = ROLLING;
+            }
+            break;
+        default:
+            break;
```

Comment:
I see some problems here. Firstly, the ball should be able to transition from any state to kicking as also happens during the game, which is currently not the case. Basically, if the velocity ever increases we need to assume the ball is kicked. 

Secondly, I am personally not a fan of the 'kicking' state. The fact the ball needs to accelerate is quite relevant and determining the initial kick velocity is a huge part of the issue here. Now if I lower threshHold kick the initial velocity is much lower than before but might not be so accurate.

I've been thinking about it and we could do some other fancy things such as formulate a linear system (similarly to how we could for chipping , but much more simple) which has positions x,y as input and the variables of initialKickVelocity, switchratio and the relevant decelleration values, and then pass a sequence of observations to it when relevant. This would also be useful in determining the actual real life variables as you would just be able to test them on a field and tune the parameters accordingly.

Reply:
As soon as the first time the kick threshold is reached, the state changes to kicking. In this state the kick velocity will be updated until the velocity reaches its maximum, this is then considered the kicking velocity. 

I agree with you that it could be extended and improved with fancy stuff.